cmd/gocognit/main.go

```diff
@@ -119,15 +133,41 @@ func analyzeDir(dirname string, stats []gocognit.Stat) []gocognit.Stat {
 }
 
 func writeStats(w io.Writer, sortedStats []gocognit.Stat) int {
-	for i, stat := range sortedStats {
-		if i == *top {
-			return i
-		}
-		if stat.Complexity <= *over {
-			return i
-		}
-		fmt.Fprintln(w, stat)
+	filter := gocognit.Filter{}
+	if *top != defaultValueIndicator {
+		// top filter
+		filter.AddFilter(gocognit.NewTopFilter(*top))
+	}
+
+	if *over != defaultValueIndicator {
+		// over filter
+		filter.AddFilter(gocognit.NewComplexityFilter(*over))
 	}
+
+	var formatter gocognit.Formatter
+
+	switch *format {
+	case textFormat:
+		formatter = gocognit.NewTextFormatter(w)
+		break
+	case jsonFormat:
+		formatter = gocognit.NewJsonFormatter(w, false)
+		break
+	case jsonPrettyFormat:
+		formatter = gocognit.NewJsonFormatter(w, true)
+		break
+	default:
+		fmt.Printf("Format '%s' is not valid, use a supported format %v", *format, supportedFormats)
+		os.Exit(1)
+	}
+
+	filtered := filter.Apply(sortedStats)
+
+	err := formatter.Write(filtered)
+	if err != nil {
+		panic(err)
+	}
+
 	return len(sortedStats)
```

Comment:
This should returned the filtered stats? Initially the return value used on `main()` to indicate wether we have stats written or not.

With your changes, the exit code not only decided  in `main()` function, but also in `writeStats(..)`. Do you have any idea to make it better?

Reply:
Yes, I've now refactored it a bit. Nevertheless this should be a very rare case.